### PR TITLE
filtering by name and price

### PR DESCRIPTION
**Pull request by ramon-reichert**
I did not touch database_test.go yet. I'll do it if you aprove what I've done until now.

### CODE REVIEW

File: cmd/api/main.go [hunk 1]
```diff
@@ -169,7 +170,50 @@ func createBook(w http.ResponseWriter, r *http.Request) {
 
 /* Returns a list of the stored books. */
 func getBooks(w http.ResponseWriter, r *http.Request) {
-	returnedBooks, err := listBooks()
+
+	err := r.ParseForm() //Verify errors in request query params and populates r.Form with its values.
```
Comment (cesarFuhr) [inline cmd/api/main.go]: This should be used for other types of requests (the ones with the `x-www-form-urlencoded` application type header).
Please use `r.URL.Query()`, it the correct one for this case https://pkg.go.dev/net/url#example-URL.Query.

Comment (cesarFuhr) [reply cmd/api/main.go]: https://go.dev/play/p/HOCbBF8c9UI

File: cmd/api/main.go [hunk 2]
```diff
@@ -169,7 +170,50 @@ func createBook(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		log.Println(err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	//Adapt query params:
+	name := r.FormValue("name")
+	if name == "" {
+		name = "%"
```
Comment (cesarFuhr) [inline cmd/api/main.go]: This is leaking SQL logic into the request handling layer, `%` is SQL specific and not `http` related.
Its better to pass the empty string to the database function and let it decide what that should be in the SQL statement.

File: cmd/api/main.go [hunk 3]
```diff
@@ -169,7 +170,50 @@ func createBook(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	var minPrice32 float32
+	minPriceStr := r.FormValue("min_price")
+	if minPriceStr != "" {
+		minPrice64, err := strconv.ParseFloat(minPriceStr, 32)
+		if err != nil {
+			log.Println(err)
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (cesarFuhr) [inline cmd/api/main.go]: This is probably a user error, if the user passes an invalid float string this error would occur.

File: cmd/api/main.go [hunk 4]
```diff
@@ -169,7 +170,50 @@ func createBook(w http.ResponseWriter, r *http.Request) {
...
+		minPrice32 = 0
+	}
+
+	var maxPrice32 float32
+	maxPriceStr := r.FormValue("max_price")
+	if maxPriceStr != "" {
+		maxPrice64, err := strconv.ParseFloat(maxPriceStr, 32)
+		if err != nil {
+			log.Println(err)
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (cesarFuhr) [inline cmd/api/main.go]: Same here.

**Review by cesarFuhr — CHANGES_REQUESTED** [review]
After these changes please add tests.
We will improve it more after the tests are written, this a good development flow:
1 - Make it work
2 - Write the tests to make sure its working as expected
3 - Refactor the code to make it easier to read/faster

**Comment by ramon-reichert** [comment]
Sorry, I've forgotten to write the test filtering by price. I'll do it now.

File: cmd/api/database_test.go [hunk 5]
```diff
@@ -97,40 +97,71 @@ func TestListBooks(t *testing.T) {
...
+	})
+
+	t.Run("List books without errors filtering by maximum price", func(t *testing.T) {
+		is := is.New(t)
+
+		//Asking all books on the created list with price <= 501
+		returnedBooks, err := listBooks("", 00.00, 501.00)
 		is.NoErr(err)
-		is.Equal(returnedBooks, bookslist)
+		is.Equal(returnedBooks[:], testBookslist[0:6])
```
Comment (cesarFuhr) [inline cmd/api/database_test.go]:
```suggestion
		is.Equal(returnedBooks, testBookslist[0:6])

File: cmd/api/database_test.go [hunk 6]
```diff
@@ -97,40 +97,71 @@ func TestListBooks(t *testing.T) {
...
+		}
+	})
+
+	t.Run("List books without errors filtering by minimum price", func(t *testing.T) {
+		is := is.New(t)
+
+		//Asking all books on the created list with price >= 501
+		returnedBooks, err := listBooks("", 501.00, 9999.99)
+		is.NoErr(err)
+		is.Equal(returnedBooks[:], testBookslist[5:11])
```
Comment (cesarFuhr) [inline cmd/api/database_test.go]:
```suggestion
		is.Equal(returnedBooks, testBookslist[5:11])

File: cmd/api/database_test.go [hunk 7]
```diff
@@ -97,40 +97,71 @@ func TestListBooks(t *testing.T) {
...
+				Price:     toPointer(float32((i * 100) + 1)),
+				Inventory: toPointer(i + 1),
 			}
 
 			newBook, err := storeOnDB(b)
 			is.NoErr(err)
 			is.Equal(newBook, b)
-			bookslist = append(bookslist, b)
+			testBookslist = append(testBookslist, b)
 		}
```
Comment (cesarFuhr) [inline cmd/api/database_test.go]: All the following tests depend on this `for` loop that creates the books to be listed.
This means that the other tests depend on this test to run correctly, lets move this `for` loop to line `111`, that way you set up all the tests in one spot, outside of the `t.Run` function.

**Review by cesarFuhr — APPROVED** [review]
Other than those little changes, LGTM.

**Review by cesarFuhr — COMMENTED** [review]
There is only one last test missing, filter by name with partial match.

Lets say you have a book called `Book of Frogs` in the database. If you filter the search based on `name`, with the filter value `Frog`, this book should be returned.

**Pull request merged**